Bugfix: Do not disconnect Koha DB handle in ReportServices controller

nugged opened the pull request:
PROBABLY that happened because some cron (non-plack) code was copied.

Or (if) .pm expected to be used from both environments, then we should move disconnects from DB to .pl / cron file only, if THAT needed.

nugged commented:
I suspect, though, this is not FULL fix for Koha-Suomi infrastructure, where 

    $dbh = C4::KohaSuomi::Tweaks->dbh();

happened, because then this connect will be to other DB as I see (like, slave, spare), and then DB needs to be very custom and disconnected. ... huh.

nugged commented:
Then we can use "C4::KohaSuomi::Tweaks" flag variable to reflect and keep disconnects only in those cases.
I added update proposal with it and rewrote the fix branch.

- - -

Yet, I'd say, it's not the best design where lifecycle for dbh inside _controller_ and moreover, in shared .pm.

nugged commented:
It is also unclear, and I don't know Koha-Suomi tricks, to you need to CLOSE ever even

    $dbh = C4::KohaSuomi::Tweaks->dbh();
    
because if yours `C4::KohaSuomi::Tweaks->dbh();` returns also global descriptor (and not reconnects) and you returns normal connection which expected to persist while plack child lasts, you will get the same problem "DBD::mysql::db prepare failed: Statement not active" on Koha-Suomi server.

nugged commented:
ah. Anyway, 

```
my $dbh;

if (try_load($module)) {
    warn "ReportServices C4::KohaSuomi::Tweaks loaded\n";
    $dbh = C4::KohaSuomi::Tweaks->dbh();

}
else {
    warn "ReportServices C4::KohaSuomi::Tweaks not loaded\n";
    $dbh = C4::Context->dbh();
}
```

executed only once per load, so anyway if we even getting **dbh** from `C4::KohaSuomi::Tweaks->dbh();` it happens once and if you under plack – you shouldn't `->disconnect` no way never.

So, probably just https://github.com/KohaSuomi/koha-plugin-report-services/commit/9cb4dc94a4fe100faf4867b4a9d147b82c899c23 might be more relevant.

johannaraisa commented:
> ah. Anyway,
> 
> ```
> my $dbh;
> 
> if (try_load($module)) {
>     warn "ReportServices C4::KohaSuomi::Tweaks loaded\n";
>     $dbh = C4::KohaSuomi::Tweaks->dbh();
> 
> }
> else {
>     warn "ReportServices C4::KohaSuomi::Tweaks not loaded\n";
>     $dbh = C4::Context->dbh();
> }
> ```
> 
> executed only once per load, so anyway if we even getting **dbh** from `C4::KohaSuomi::Tweaks->dbh();` it happens once and if you under plack – you shouldn't `->disconnect` no way never.
> 
> So, probably just [9cb4dc9](https://github.com/KohaSuomi/koha-plugin-report-services/commit/9cb4dc94a4fe100faf4867b4a9d147b82c899c23) might be more relevant.

You are right that `->disconnect` isn't needed there. Can you update the pull request so it is removed?

nugged committed:
Bugfix: Do not disconnect Koha DB handle in ReportServices controller

Koha's database handle from C4::Context/Koha::Database is shared
and managed per Plack worker. Calling $dbh->disconnect() inside
getReportData() randomly breaks that shared connection for the
whole worker.

As a result, the first call to
/api/v1/contrib/kohasuomi/reportservices/reports works, but
subsequent calls in the same worker fail with:

  DBIx::Class::Exception:
    DBD::mysql::db prepare failed: Statement not active

Because the same Plack worker continues serving other requests
for possibly different logged-in users, this forced disconnect
can also make unrelated Koha pages fail randomly with the same
database error until the worker is recycled or have failover reconnect.

Remove the explicit $dbh->disconnect() calls from both the normal
and exception paths and let Koha manage the connection lifecycle.

nugged commented:
re-published simple removal.
